gslib/utils/shim_util.py

```diff
@@ -0,0 +1,219 @@
+    return args + gsutil_args
+
+  def get_gcloud_storage_args(self):
+    """Translates the gsutil command flags to gcloud storage flags.
+
+    It uses the command_spec.gcloud_storage_map field that provides the
+    translation mapping for all the flags.
+    
+    Returns:
+      A list of all the options and arguments that can be used with the
+        equivalent gcloud storage command.
+    Raises:
+      GcloudStorageTranslationError: If a flag or command cannot be translated.
+      ValueError: If there is any issue with the mapping provided by
+        GcloudStorageMap.
+    """
+    return self._get_gcloud_storage_args(self.sub_opts, self.args,
+                                         self.gcloud_storage_map)
+
+  def _print_gcloud_storage_command_info(self,
+                                         gcloud_command,
+                                         env_variables,
+                                         dry_run=False):
+    logger_func = self.logger.info if dry_run else self.logger.debug
+    logger_func('Gcloud Storage Command: {}'.format(' '.join(gcloud_command)))
+    if env_variables:
+      logger_func('Enviornment variables for Gcloud Storage:')
+      for k, v in env_variables.items():
+        logger_func('%s=%s', k, v)
+
+  def translate_to_gcloud_storage_if_requested(self):
+    """Translates the gsutil command to gcloud storage equivalent.
+
+    The translated commands get stored at
+    self._translated_gcloud_storage_command.
+    This command also translate the boto config, which gets stored as a dict
+    at self._translated_env_variables
+    
+    Returns:
+      True if the command was successfully translated, else False.
+    """
+    use_gcloud_storage = config.get('GSUtil', 'use_gcloud_storage', 'never')
+    if use_gcloud_storage not in VALID_USE_GCLOUD_STORAGE_VALUES:
+      raise exception.CommandException(
+          'Invalid option specified for'
+          ' GSUtil:use_gcloud_storage config setting. Should be one of: {}'.
+          format(' | '.join(VALID_USE_GCLOUD_STORAGE_VALUES)))
+    if use_gcloud_storage != 'never':
+      try:
+        # TODO(b/206143429) Get top level flags.
+        top_level_flags = []
+
+        gcloud_binary_path = _get_gcloud_binary_path()
+        gcloud_storage_command = ([gcloud_binary_path] +
+                                  self.get_gcloud_storage_args() +
+                                  top_level_flags)
+        # TODO(b/206149936): Translate boto config to CLOUDSDK envs.
+        env_variables = {}
+        if use_gcloud_storage == 'dry_run':
+          self._print_gcloud_storage_command_info(gcloud_storage_command,
+                                                  env_variables,
+                                                  dry_run=True)
+        elif not os.environ.get('CLOUDSDK_CORE_PASS_CREDENTIALS_TO_GSUTIL'):
+          raise exception.GcloudStorageTranslationError(
+              'Gsutil is not using the same credentials as gcloud.'
```

Comment:
Should we mention this is only an issue b/c the shim is active?

Reply:
Good point. Done.